File: 329-longest-increasing-path-in-a-matrix/329-longest-increasing-path-in-a-matrix.java
class Solution {
    
    int[][] directions = {{1,0},{-1,0},{0,1},{0,-1}};
    
    public int longestIncreasingPath(int[][] matrix) {
        
        int m = matrix.length, n = matrix[0].length;
        int max = Integer.MIN_VALUE;
        int[][] dp = new int[m][n];
        for(int i  = 0 ; i < m ; i++){
            for(int j = 0; j < n; j++){
                max = Math.max(dfs(i,j,dp,matrix),max);
            }
        }
        
        return max + 1;
    }
    
    private int dfs(int i, int j, int[][] v, int[][] matrix){
        if(outBound(i,j,matrix.length,matrix[0].length))
            return 0;
        if(v[i][j] != 0)
            return v[i][j];
        int res = 0;
        for(int[] dir : directions){
            int x = i + dir[0], y = j + dir[1];
            if(!outBound(x,y,matrix.length,matrix[0].length) && matrix[i][j] < matrix[x][y])
                res = Math.max(res,1 + dfs(x,y,v,matrix));
        }
        return v[i][j] = res;
    }
    
    private boolean outBound(int i ,int j , int m , int n){
        return i < 0 || j < 0 || i >= m || j >= n;
    }
}

File: 329-longest-increasing-path-in-a-matrix/329-longest-increasing-path-in-a-matrix.py
class Solution:
    def longestIncreasingPath(self, matrix: List[List[int]]) -> int:
        n = len(matrix)
        m = len(matrix[0])
        
        dp = [[None for _ in range(m)] for _ in range(n)]
        
        dirs = [(0,1),(0,-1),(1,0),(-1,0)]
        def dfs(i,j):
            #print(i,j)
            #print("f", dp)
            if dp[i][j] != None:
                return dp[i][j]
                
            res = 1
            for adj in dirs:
                x,y = adj[0] + i, adj[1] + j
                if 0 <= x < n and 0 <= y < m and matrix[i][j] < matrix[x][y]:
                    res = max(res,1 + dfs(x,y))
            dp[i][j] = res
            #print("l",dp)
            return res
        
        ans = 0
        
        for i in range(n):
            for j in range(m):
                ans = max(ans,dfs(i,j))
        return ans